### PR TITLE
Add ios_logging_global module 

### PR DESCRIPTION
**Pull request by KB-perByte**
##### SUMMARY

Resource module for ios logging - cisco.ios_logging_global
##### ISSUE TYPE

- New Module Pull Request

##### COMPONENT NAME
ios_logging_global

### CODE REVIEW

**Comment by KB-perByte** [comment]
recheck

File: tests/unit/modules/network/ios/test_ios_logging_global.py [hunk 1]
```diff
@@ -0,0 +1,766 @@
+#
+# (c) 2019, Ansible by Red Hat, inc
```
Comment (ashwini-mhatre) [inline tests/unit/modules/network/ios/test_ios_logging_global.py:2]: this should be 2021

File: plugins/modules/ios_logging_global.py [hunk 2]
```diff
@@ -0,0 +1,1060 @@
...
+          - hostname: 172.16.1.10
+            filtered: true
+            stream: 15
+    state: replaced
+
+# Commands Fired:
+# ---------------
+
+# "commands": [
+#         "no logging host 172.16.1.13",
```
Comment (justjais) [inline plugins/modules/ios_logging_global.py:655]: @KB-perByte like we discussed with this, what's the difference b/w `overridden` and `replace` state, ideally imo for list of dict params `replace` should replace content of only 1 config param unlike `overridden` state which will effect all of the params of list of dict, like for config scenario:
```
logging persistent threshold 2
logging facility local6
logging host 172.16.0.1 filtered session-id string test
logging host 10.55.240.143
logging host 10.0.2.15 transport tcp audit
logging host 172.16.0.1 transport tcp
logging host 172.16.0.2
```
Play:
```
 name: Replace commands with provided configuration
  cisco.ios.ios_logging_global:
    config:
      - hosts:
          - hostname: 172.16.0.1
             sequence-num-session: true
```
Output for replace should be:
```
[
  "no logging host 172.16.0.1 filtered session-id string test",
  "no logging host 172.16.0.1 transport tcp",
  "logging host 172.16.0.1 sequence-num-session"
]
```
Output for override:
```
[
  "no logging host 10.55.240.143",
  "no logging host 172.16.0.2",
  "no logging host 172.16.0.1 filtered session-id string test",
  "no logging host 172.16.0.1 transport tcp",
  "logging host 172.16.0.1 sequence-num-session"
]
```

File: plugins/modules/ios_logging_global.py [hunk 3]
```diff
@@ -0,0 +1,1060 @@
...
+# logging snmp-trap errors
+# logging host 172.16.1.13 transport tcp port 514
+# logging host 172.16.1.11 xml
+# logging host 172.16.1.12
+# logging host 172.16.1.10 filtered stream 10
+# logging host 172.16.1.25 filtered
+
+- name: Gather listed logging config
+  cisco.ios.ios_logging_global:
+    config:
```
Comment (justjais) [inline plugins/modules/ios_logging_global.py:713]: u can remove `config` from the play

**Review by justjais — CHANGES_REQUESTED** [review]

File: plugins/modules/ios_logging_global.py [hunk 4]
```diff
@@ -0,0 +1,1060 @@
...
+# logging host 172.16.1.13 transport tcp port 514
+# logging host 172.16.1.11 xml
+# logging host 172.16.1.12
+# logging host 172.16.1.10 filtered stream 10
+
+- name: Remove as per the provided configuration
+  cisco.ios.ios_logging_global:
+    config:
+      - snmp_trap:
+          - errors
```
[the remark this answers is not in the document]
Comment (KB-perByte) [reply plugins/modules/ios_logging_global.py:508]: yes, added two elements as operations to give the idea that It would generate two commands

File: plugins/modules/ios_logging_global.py [hunk 5]
```diff
@@ -0,0 +1,1060 @@
...
+        description: Enable logging of user info on privileged mode enabling
+        type: bool
+  running_config:
+      description:
+      - This option is used only with state I(parsed).
+      - The value of this option should be the output received from the IOS device by
+        executing the command B(show running-config | include logging).
+      - The state I(parsed) reads the configuration from C(running_config) option and
+        transforms it into Ansible structured data as per the resource module's argspec
+        and the value is then returned in the I(parsed) key within the result.
```
[the remark this answers is not in the document]
Comment (KB-perByte) [reply plugins/modules/ios_logging_global.py:360]: Where should be the right place for this information? running_config is specifically for the parsed state right!

File: plugins/modules/ios_logging_global.py (hunk 2 above)
Comment (KB-perByte) [reply plugins/modules/ios_logging_global.py:655]: As per the discussion we had in the calls/ demo 
**Overridden** - to override the whole logging configuration and put in what is there in the playbook and negating everything else.
Whereas,
**Replaced** - to replace the whole list of **have** configuration for host / other list params individually.
**Tagging in examples-** 
```
Existing running-config -
logging persistent threshold 2
logging facility local6
logging host 172.16.0.1 filtered session-id string test
logging host 10.55.240.143
logging host 10.0.2.15 transport tcp audit
logging host 172.16.0.1 transport tcp
logging host 172.16.0.2
```
Playbook - 
```
  - name: Replace commands with provided configuration
    cisco.ios.ios_logging_global:
      config:
        - hosts: 
            - hostname: 10.0.1.19
      state: replaced
```
Commands fired -
```
no logging host 172.16.0.1
no logging host 10.55.240.143
no logging host 10.0.2.15
no logging host 172.16.0.1
no logging host 172.16.0.2
logging host 10.0.1.19
```
for a case to update a single logging host one can use the merged state.

Comment (justjais) [reply plugins/modules/ios_logging_global.py:655]: hmm.. I might have missed the discussion then, and if it was discussed and agreed upon then it's OK.

File: plugins/modules/ios_logging_global.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (justjais) [reply plugins/modules/ios_logging_global.py:360]: @KB-perByte u can document the same in `state` documentation

File: tests/integration/targets/ios_logging_global/tests/cli/deleted.yaml [hunk 6]
```diff
@@ -0,0 +1,22 @@
+---
+- debug:
+    msg: Start Deleted integration state for ios_logging ansible_connection={{
+      ansible_connection }}
+
+- include_tasks: _remove_config.yaml
+- include_tasks: _populate_config.yaml
+
+- block:
```
Comment (justjais) [inline tests/integration/targets/ios_logging_global/tests/cli/deleted.yaml:9]: can u include delete operation task for few params as well in addition to delete_all

File: tests/integration/targets/ios_logging_global/tests/cli/merged.yaml [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+              rate_limit: 10
+          - cns_events: warnings
+        state: merged
+      register: result
+
+    - name: TEST - [merged] Assert that correct set of commands were generated
+      assert:
+        that:
+          - "{{ merged['commands'] | symmetric_difference(result['commands']) |\
+            \ length == 0 }}"
```
Comment (justjais) [inline tests/integration/targets/ios_logging_global/tests/cli/merged.yaml:52]: @KB-perByte I see u're comparing cmds only, are u not comparing before and after list of dict generated.

File: tests/integration/targets/ios_logging_global/tests/cli/overridden.yaml [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+          - hosts:
+              - hostname: 172.16.1.12
+                xml: true
+        state: overridden
+
+    - name: Assert that correct set of commands were generated
+      assert:
+        that:
+          - "{{ overridden['commands'] | symmetric_difference(result['commands'])\
+            \ | length == 0 }}"
```
Comment (justjais) [inline tests/integration/targets/ios_logging_global/tests/cli/overridden.yaml:31]: same as merge comment

File: tests/integration/targets/ios_logging_global/tests/cli/replaced.yaml [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+          - facility: local6
+          - hosts:
+              - hostname: 172.16.1.19
+                filtered: false
+        state: replaced
+
+    - name: TEST - [replaced] Assert that correct set of commands were generated
+      assert:
+        that:
+          - "{{ replaced['commands'] | symmetric_difference(result['commands'])\
```
Comment (justjais) [inline tests/integration/targets/ios_logging_global/tests/cli/replaced.yaml:27]: same as merge comment

File: tests/integration/targets/ios_logging_global/tests/cli/_populate_config.yaml [hunk 10]
```diff
@@ -0,0 +1,38 @@
+---
```
Comment (justjais) [inline tests/integration/targets/ios_logging_global/tests/cli/_populate_config.yaml:1]: @KB-perByte Gathered and empty_config integration tests are missing

File: tests/integration/targets/ios_logging_global/vars/main.yaml [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+    - logging message-counter syslog
+    - logging buffered xml 5099 notifications
+    - logging console xml critical
+    - logging monitor warnings
+    - logging exception 4099
+    - logging host 172.16.1.10 filtered stream 10
+    - logging cns-events warnings
+    - logging trap errors
+
+  after:
```
Comment (justjais) [inline tests/integration/targets/ios_logging_global/vars/main.yaml:24]: after dict maintained but it's not being compared with integration tests run, ideally before and after should be compared for all merge, replace and override state.

File: tests/unit/modules/network/ios/test_ios_logging_global.py [hunk 12]
```diff
@@ -0,0 +1,766 @@
...
+            "no logging filter flash:172.16.1.1 1 args Test",
+        ]
+        playbook["state"] = "deleted"
+        set_module_args(playbook)
+        result = self.execute_module(changed=True)
+
+        self.maxDiff = None
+        self.assertEqual(sorted(result["commands"]), sorted(deleted))
+
+    def test_ios_logging_global_replaced(self):
```
Comment (justjais) [inline tests/unit/modules/network/ios/test_ios_logging_global.py:739]: `test_ios_logging_global_replaced_idempotent` UTC is missing

File: tests/unit/modules/network/ios/test_ios_logging_global.py [hunk 13]
```diff
@@ -0,0 +1,766 @@
...
+            "no logging source-interface GBit1/0",
+            "no logging source-interface CTunnel2",
+        ]
+        playbook["state"] = "deleted"
+        set_module_args(playbook)
+        result = self.execute_module(changed=True)
+        self.maxDiff = None
+        self.assertEqual(sorted(result["commands"]), sorted(deleted))
+
+    def test_ios_logging_global_overridden(self):
```
Comment (justjais) [inline tests/unit/modules/network/ios/test_ios_logging_global.py:391]: `test_ios_logging_global_overridden_idempotent` UTC is missing

File: tests/unit/modules/network/ios/test_ios_logging_global.py [hunk 14]
```diff
@@ -0,0 +1,766 @@
...
+        )
+        merged = []
+        playbook["state"] = "merged"
+        set_module_args(playbook)
+        result = self.execute_module()
+
+        self.maxDiff = None
+        self.assertEqual(sorted(result["commands"]), sorted(merged))
+
+    def test_ios_logging_global_deleted(self):
```
Comment (justjais) [inline tests/unit/modules/network/ios/test_ios_logging_global.py:226]: `test_ios_logging_global_deleted_idempotent` scenario is missing

**Review by justjais — CHANGES_REQUESTED** [review]

**Comment by KB-perByte** [comment]
recheck

File: tests/integration/targets/ios_facts/tests/cli/all_facts.yaml [hunk 15]
```diff
@@ -6,7 +6,7 @@
   cisco.ios.ios_facts:
     gather_subset:
       - all
-    provider: '{{ cli }}'
```
Comment (justjais) [inline tests/integration/targets/ios_facts/tests/cli/all_facts.yaml:9]: why is this file updated, as I see there's no change here wrt to the module leaving the addition of module name under `set_fact`, if not intended can u revert the change.

**Review by justjais — COMMENTED** [review]
@KB-perByte changes LGTM, there's 1 cosmetic comment. I'm giving +1, plz address the comment before gating the PR

**Review by justjais — APPROVED** [review]

File: tests/integration/targets/ios_facts/tests/cli/all_facts.yaml (hunk 15 above)
Comment (KB-perByte) [reply tests/integration/targets/ios_facts/tests/cli/all_facts.yaml:9]: That is the tox scan was reformatting it, let me revert the changes.

Comment (KB-perByte) [reply tests/integration/targets/ios_facts/tests/cli/all_facts.yaml:9]: tox is updating the formatting, and the files changed produces warnings or gets reformatted on tox scan. 
As it is not making any of the tests fail. Gate'ing it.

**Comment by KB-perByte** [comment]
recheck

**Comment by KB-perByte** [comment]
recheck

**Comment by KB-perByte** [comment]
recheck

**Review by ashwini-mhatre — APPROVED** [review]

**Comment by KB-perByte** [comment]
recheck
